ext/json/json_encoder.c

```diff
@@ -528,6 +528,26 @@ static int php_json_encode_serializable_object(smart_str *buf, zval *val, int op
 }
 /* }}} */
 
+static int php_json_encode_serializable_enum(smart_str *buf, zval *val, int options, php_json_encoder *encoder)
+{
+	zend_class_entry *ce = Z_OBJCE_P(val);
+	if (ce->enum_scalar_type == IS_UNDEF) {
+		encoder->error_code = PHP_JSON_ERROR_NON_SCALAR_ENUM;
+		smart_str_appendc(buf, '0');
+		return FAILURE;
+	}
+
+	zend_string *value_string = zend_string_init("value", strlen("value"), 0);
+	zval tmp;
+	zval *value_zv = zend_read_property_ex(ce, Z_OBJ_P(val), value_string, 0, &tmp);
+
+	int result = php_json_encode_zval(buf, value_zv, options, encoder);
+
+	zend_string_free(value_string);
+
+	return result;
+}
```

Comment:
Not sure it's a good idea, but wanted to mention it anyway: We could also automatically implement JsonSerializable for backed enums.

Reply:
This could cause issues with traits I think. As the traits are bound at run time we wouldn't know if we need to provide a default implementation, but the default implementation will have precedence over the trait implementation.